examples/www-image/main/jpg-render.c

```diff
@@ -0,0 +1,518 @@
+};
+
+const uint16_t ep_width = EPD_WIDTH;
+const uint16_t ep_height = EPD_HEIGHT;
+uint8_t gamme_curve[256];
+
+static const char *TAG = "EPDiy";
+uint16_t countDataEventCalls = 0;
+uint32_t countDataBytes = 0;
+uint32_t img_buf_pos = 0;
+uint32_t dataLenTotal = 0;
+uint64_t startTime = 0;
+
+/*====================================================================================
+  This sketch contains support functions to render the Jpeg images.
+
+  Created by Bodmer 15th Jan 2017
+  ==================================================================================*/
+
+// Return the minimum of two values a and b
+#define minimum(a,b)     (((a) < (b)) ? (a) : (b))
+
+uint8_t find_closest_palette_color(uint8_t oldpixel)
+{
+  return (round((oldpixel / 16)*16));
+}
+
+//====================================================================================
+//   Decode and paint onto the Epaper screen
+//====================================================================================
+void jpegRender(int xpos, int ypos, int width, int height) {
+ #if JPG_DITHERING
+ unsigned long pixel=0;
+ for (uint16_t by=0; by<ep_height;by++)
+  {
+    for (uint16_t bx=0; bx<ep_width;bx++)
+    {
+        int oldpixel = decoded_image[pixel];
+        int newpixel = find_closest_palette_color(oldpixel);
+        int quant_error = oldpixel - newpixel;
+        decoded_image[pixel]=newpixel;
+        if (bx<(ep_width-1))
+          decoded_image[pixel+1] = minimum(255,decoded_image[pixel+1] + quant_error * 7 / 16);
+
+        if (by<(ep_height-1))
+        {
+          if (bx>0)
+            decoded_image[pixel+ep_width-1] =  minimum(255,decoded_image[pixel+ep_width-1] + quant_error * 3 / 16);
+
+          decoded_image[pixel+ep_width] =  minimum(255,decoded_image[pixel+ep_width] + quant_error * 5 / 16);
+          if (bx<(ep_width-1))
+            decoded_image[pixel+ep_width+1] = minimum(255,decoded_image[pixel+ep_width+1] + quant_error * 1 / 16);
+        }
+        pixel++;
+    }
+  }
+  #endif
+
+  // Write to display
+  uint64_t drawTime = esp_timer_get_time();
+  uint32_t padding_x = (EPD_WIDTH - width) / 2;
+  uint32_t padding_y = (EPD_HEIGHT - height) / 2;
+
+  for (uint32_t by=0; by<height;by++) {
+    for (uint32_t bx=0; bx<width;bx++) {
+        epd_draw_pixel(bx + padding_x, by + padding_y, decoded_image[by * width + bx], fb);
+    }
+  }
+  // calculate how long it took to draw the image
+  time_render = (esp_timer_get_time() - drawTime)/1000;
+  ESP_LOGI("render", "%d ms - jpeg draw", time_render);
+}
+
+void deepsleep(){
+    esp_deep_sleep(1000000LL * 60 * DEEPSLEEP_MINUTES_AFTER_RENDER);
+}
+
+static uint32_t feed_buffer(JDEC *jd,      
+               uint8_t *buff, // Pointer to the read buffer (NULL:skip) 
+               uint32_t nd 
+) {
+    vTaskDelay(1);
+    esp_task_wdt_reset();
+    uint32_t count = 0;
+
+    while (count < nd) {
+      if (buff != NULL) {
+            uint8_t b = source_buf[buffer_pos];
```

Comment:
We could make source_buf much smaller by decompressing while the data comes in, but we'd need to keep track of how much was read / written. Are you interested in adding this?